### PR TITLE
Add a new machine model for flexflow simulator

### PR DESCRIPTION
**Pull request by dycz0fx**
1. Restructured the simulator's code. Move the current code that describes the status of a machine into a separate class (MachineModel_old). This machine model would be used by default, and the simulator should behave the same as before.
2. Implemented a new machine model (MachineModel_new), which supports creating a machine model from a configure file (please see macine_config_example), and simulating message segmentation and possible congestions on a communication device.

### CODE REVIEW

File: include/simulator.h
Comment (lockshaw) [inline include/simulator.h:69]:
```suggestion
class MemDevice : public Device {
```

File: include/simulator.h
Comment (lockshaw) [inline include/simulator.h:58]:
```suggestion
class CompDevice : public Device {
```

File: include/simulator.h
Comment (lockshaw) [inline include/simulator.h:112]:
```suggestion
class SimpleMachineModel : public MachineModel {
```

File: include/simulator.h
Comment (lockshaw) [inline include/simulator.h:96]:
```suggestion
    Comm_device(std::string const &name, CommDevType comm_type, int node_id, int socket_id, int device_id, float latency, float bandwidth);
```

File: include/simulator.h
Comment (lockshaw) [inline include/simulator.h:130]:
```suggestion
  std::unordered_map<int, Comp_device*> id_to_gpu;
```

File: include/simulator.h
Comment (lockshaw) [inline include/simulator.h:138]:
```suggestion
class FullMachineModel : public MachineModel {
```

File: include/simulator.h
Comment (lockshaw) [inline include/simulator.h:193]: Would it be reasonable to use a `std::unordered_map` with a struct key that makes these two indices clear, or is there a compelling reason for nested vectors?

File: include/simulator.h
Comment (lockshaw) [inline include/simulator.h:308]:
```suggestion
  SimTask* new_comm_task(std::string const &name, Comm_device *comm_device, size_t message_size);
```

File: include/simulator.h
Comment (lockshaw) [inline include/simulator.h:114]: Feels like a lot of these methods could/should be `const`

File: include/simulator.h
Comment (lockshaw) [inline include/simulator.h:61]: Why are the names here unintuitive and require comments?

File: src/runtime/simulator.cu
Comment (lockshaw) [inline src/runtime/simulator.cu:91]: More useful error message here since this is not unlikely to happen?

File: src/runtime/simulator.cc
Comment (lockshaw) [inline src/runtime/simulator.cc:601]: Can we use iterators or something to make this code a little easier to read?

File: src/runtime/simulator.cc
Comment (lockshaw) [inline src/runtime/simulator.cc:571]:
```suggestion
  for (int node_id = 0; node_id < num_nodes; node_id++) {
```

File: src/runtime/simulator.cc
Comment (lockshaw) [inline src/runtime/simulator.cc:1099]: Divide this out into a couple expressions. It's quite hard to read

File: src/runtime/simulator.cc
Comment (lockshaw) [inline src/runtime/simulator.cc:185]: Can we replace a bunch of these places where we do the hashing explicitly with instead creating structs and providing them `std::hash` instances? The inline hashing code is bug-prone and makes things hard to read

File: src/runtime/simulator.cc
Comment (lockshaw) [inline src/runtime/simulator.cc:201]: Should this `exit` or `assert` or something, or is the expected behavior to just return an empty list?

File: src/runtime/simulator.cc
Comment (lockshaw) [inline src/runtime/simulator.cc:407]: I'd personally prefer `this->` in front of these to make clear that they're methods and not functions, but this is a small thing. I defer to @jiazhihao on this

File: src/runtime/simulator.cc
Comment (lockshaw) [reply src/runtime/simulator.cc:571]: In general if we can substitute out most of the explicit indices with real names the code would be much easier to read

**Review by lockshaw — CHANGES_REQUESTED** [review]
Overall it looks great, though I haven't so far been able to fully understand all of the improvements to the old simulator. I left a bunch of comments mostly focused on cleaning up some of the code and making it a bit easier to read. In general if if something was recurring I only commented once, so many of the suggestions should ideally be applied throughout the file. That said, I'm absolutely happy to discuss any suggestions you think are wrong :slightly_smiling_face:! Once you/we've worked through these, let me know and I'd love to give everything another read through before merging it. Thanks!

A couple comments that didn't make sense to put at a specific line of code: 
- Would it be possible to split some of the new machine model code out of simulator.cc? The file is getting quite large and it would be nice to have a clearer division between what code is related to the new machine model and what is related to the old one.
- It would also be great to have some doxygen-style docstrings provided for some of the more high-level functions in the new simulator, as well as a plain-text description summarizing roughly what the model includes and what it doesn't

File: include/simulator.h
Comment (dycz0fx) [reply include/simulator.h:216]: Fixed

File: src/runtime/simulator.cc
Comment (dycz0fx) [reply src/runtime/simulator.cc:718]: I have an impression that emplace_back can save a copy as compared to push_back.

File: src/runtime/simulator.cc
Comment (dycz0fx) [reply src/runtime/simulator.cc:697]: fixed

File: src/runtime/simulator.cc
Comment (dycz0fx) [reply src/runtime/simulator.cc:931]: fixed

File: src/runtime/simulator.cc
Comment (dycz0fx) [reply src/runtime/simulator.cc:244]: I used prinf because I saw the original code use it. What kind of logging mechanism would you suggest?

File: src/runtime/simulator.cc
Comment (dycz0fx) [reply src/runtime/simulator.cc:42]: fixed

File: machine_config_example
Comment (dycz0fx) [reply machine_config_example:15]: I used ms because the calculations in the simulator are in ms, but we can change to other units.

File: src/runtime/simulator.cu
Comment (dycz0fx) [reply src/runtime/simulator.cu:85]: fixed

File: src/runtime/model.cc
Comment (dycz0fx) [reply src/runtime/model.cc:1928]: fixed

File: include/simulator.h
Comment (dycz0fx) [reply include/simulator.h:69]: fixed

File: include/simulator.h
Comment (dycz0fx) [reply include/simulator.h:58]: fixed

File: include/simulator.h
Comment (dycz0fx) [reply include/simulator.h:112]: fixed

File: include/simulator.h
Comment (dycz0fx) [reply include/simulator.h:96]: fixed

File: include/simulator.h
Comment (dycz0fx) [reply include/simulator.h:130]: I think it makes sense to change to unordered_map, but the original code uses map. Just trying to make the simple machine model be the same as the original one.

File: include/simulator.h
Comment (dycz0fx) [reply include/simulator.h:138]: Maybe EnhancedMachineModel?

File: include/simulator.h
Comment (dycz0fx) [reply include/simulator.h:308]: fixed

File: include/simulator.h
Comment (dycz0fx) [reply include/simulator.h:114]: fixed

File: include/simulator.h
Comment (dycz0fx) [reply include/simulator.h:61]: Just using Legion's name for the CPU and GPU.

File: src/runtime/simulator.cc
Comment (dycz0fx) [reply src/runtime/simulator.cc:201]: fixed

File: src/runtime/simulator.cc
Comment (dycz0fx) [reply src/runtime/simulator.cc:407]: fixed

File: src/runtime/simulator.cc
Comment (dycz0fx) [reply src/runtime/simulator.cc:185]: Make sense. When implementing the simple machine model, I tried to make it be the same as the original code, so the default behavior of Flexflow wouldn't be changed.

File: src/runtime/simulator.cc
Comment (dycz0fx) [reply src/runtime/simulator.cc]: I did this way because I'd like to make this function virtual. It is a little bit trickier to make operator << virtual.

File: machine_config_example
Comment (dycz0fx) [reply machine_config_example:1]: I feel like the custom configuration file is pretty small and easy to handle. Using YMAL or XML is more standardized, but could make Flexflow require more external packages.

File: src/runtime/simulator.cu
Comment (dycz0fx) [reply src/runtime/simulator.cu:91]: Fixed.

**Comment by dycz0fx** [comment]
Thanks for reviewing the pull request! 
I addressed most of the comments and please take another look. Some comments about optimizing the old machine model make sense, but I feel like those should not be included in this pull request.

**Pull request merged**
